src/helpers/binding.ts

```diff
@@ -0,0 +1,163 @@
+  return Number(param).toString();
+}
+
+function bindNumber(param: number | bigint): string {
+  return param.toString();
+}
+
+function bindDate(param: Date): string {
+  return param.valueOf().toString();
+}
+
+function bindBytes(param: Uint8Array): string {
+  const hex = param.reduce((t, x) => t + x.toString(16).padStart(2, "0"), "");
+  return `X'${hex}'`;
+}
+
+function bindObject(param: any): string {
+  return bindString(JSON.stringify(param));
+}
+
+function bindNull(_param: undefined | null): string {
+  return "NULL";
+}
+
+function bindToString(param: any): string {
+  return bindString(String(param));
+}
+
+interface SQL {
+  toSQL: () => string;
+}
+
+function isSQL(param: any): param is SQL {
+  return typeof param.toSQL === "function";
+}
+
+function bindToSQL(param: SQL): string {
+  return param.toSQL();
+}
+
+function bindValue(param: BaseType): string {
+  switch (typeof param) {
+    case "bigint":
+    case "number":
+      return bindNumber(param);
+    case "boolean":
+      return bindBoolean(param);
+    case "string":
+      return bindString(param);
+    case "undefined":
+      return bindNull(param);
+    case "object":
+      if (param instanceof Date) {
+        return bindDate(param);
+      } else if (param instanceof Uint8Array) {
+        return bindBytes(param);
+      } else if (param == null) {
+        return bindNull(param);
+      } else if (isPlainObject(param)) {
+        return bindObject(param);
+      } else if (isSQL(param)) {
+        return bindToSQL(param);
+        /* c8 ignore next 3 */
+      } else {
+        return bindToString(param);
+      }
+    default:
+      return bindToString(param);
+  }
+}
+
+export type BaseType =
+  | string // strings are left as is (and escaped)
+  | boolean // boolean is converted to ints
+  | number // numbers are left as is
+  | bigint // bigints are converted to ints
+  | Uint8Array // bytes arrays are converted to byte strings
+  | null // null is converted to NULL
+  | undefined // undefined is converted to NULL
+  | Date // Date objects are converted to their valueOf
+  | SQL // Anything that has a toSQL method
+  | Record<string, any>; // JSON objects
+
+export type ValuesType = BaseType | BaseType[] | Record<string, BaseType>;
+
+export interface Parameters {
+  anon: BaseType[];
+  named?: Record<string, BaseType>;
+}
+
+export function getParameters(...values: ValuesType[]): Parameters {
+  const initialValue: Required<Parameters> = { anon: [], named: {} };
+  const flat = values.flat(Infinity);
+  const result = flat.reduce(
+    ({ anon, named }: Required<Parameters>, v: any) => {
+      if (isPlainObject(v)) {
+        return { anon, named: { ...named, ...v } };
+      } else {
+        return { anon: [...anon, v], named };
+      }
+    },
+    initialValue
+  );
+  return result;
+}
+
+export function bindValues(sql: string, parameters?: Parameters): string {
+  // https://sqlite.org/forum/forumpost/4350e973ad
```

Comment:
I learned a lot about binding values and how SQLite handles all its parameter values in SQL strings.